odl/discr/discr_utils.py

```diff
@@ -0,0 +1,812 @@
+    types like strings:
+    >>> part = odl.uniform_partition(0, 3, 6)
+    >>> part.coord_vectors  # grid points
+    (array([ 0.25,  0.75,  1.25,  1.75,  2.25,  2.75]),)
+    >>> f = ['s', 't', 'r', 'i', 'n', 'g']
+    >>> interpolator = nearest_interpolator(f, part.coord_vectors)
+    >>> print(interpolator(1.0))
+    t
+    See Also
+    --------
+    linear_interpolator : (bi-/tri-/...)linear interpolation
+    per_axis_interpolator : potentially different interpolation in each axis
+    Notes
+    -----
+    - **Important:** if called on a point array, the points are
+      assumed to be sorted in ascending order in each dimension
+      for efficiency reasons.
+    - Nearest neighbor interpolation is the only scheme which works
+      with data of non-numeric data type since it does not involve any
+      arithmetic operations on the values, in contrast to other
+      interpolation methods.
+    - The distinction between left and right variants is currently
+      made by changing ``<=`` to ``<`` at one place. This difference
+      may not be noticable in some situations due to rounding errors.
+    """
+    f = np.asarray(f)
+
+    # TODO(kohr-h): pass reasonable options on to the interpolator
+    def nearest_interp(arg, out=None):
+        """Interpolating function with vectorization."""
+        if is_valid_input_meshgrid(arg, f.ndim):
+            input_type = 'meshgrid'
+        else:
+            input_type = 'array'
+
+        interpolator = _NearestInterpolator(
+            coord_vecs,
+            f,
+            variant=variant,
+            input_type=input_type
+        )
+
+        return interpolator(arg, out=out)
+
+    return nearest_interp
+
+
+def linear_interpolator(f, coord_vecs):
+    """Return the linear interpolator for discrete values.
+    Parameters
+    ----------
+    f : numpy.ndarray
+        Function values that should be interpolated.
+    coord_vecs : sequence of numpy.ndarray
+        Coordinate vectors of the rectangular grid on which interpolation
+        should be based. They must be sorted in ascending order. Usually
+        they are obtained as ``grid.coord_vectors`` from a `RectGrid`.
+    Returns
+    -------
+    interpolator : function
+        Python function that will interpolate the given values when called
+        with a point or multiple points (vectorized).
+    Examples
+    --------
+    We interpolate a 1d function. If called with a single point, the
+    interpolator returns a single value, and with multiple points at once,
+    an array of values is returned:
+    >>> part = odl.uniform_partition(0, 2, 5)
+    >>> part.coord_vectors  # grid points
+    (array([ 0.2,  0.6,  1. ,  1.4,  1.8]),)
+    >>> f = [1.0, 2.0, 3.0, 4.0, 5.0]
+    >>> interpolator = linear_interpolator(f, part.coord_vectors)
+    >>> interpolator(0.3)  # 0.75 * 1 + 0.25 * 2 = 1.25
+    1.25
+    >>> interpolator([0.6, 1.3, 1.9])
+    array([ 2.  ,  3.75,  3.75])
```

Comment:
It is hard to, from the above doc, understand why `1.9` gets interpolated to `3.75`. It must have to do with an implicit assumption of the behavior of `f` at the boundary. What is the assumption? Can the user modify this by passing extra arguments?